Bugfix/from meeting 2024 06 21

MKacprzakSolvd opened the pull request:
Things to fix:
  - [x] improve .gitignore to add support for other IDE's (get from Carina (or from github))
  - [X] create class that provides data instead of using data provider
    or add more data to data provider (to check different countries)
  - [X] add test description to all tests
  - [X] in first test with printing product names add soft assert to
  - [X] when opening products use different strategy than url (don't work for subcategories)
  - [X] make it work with bugged carina webdriver
  - [X] addRemove product card -> add emptying basket
  - [X] in checking filter add to check whether correct color / size is selected
  - [X] in isProduct in card check also that price is right
  - [X] in verifyCheckoutFromItemDetailsPageTest add comment to assert (there is assert with on)

MKacprzakSolvd committed:
added checking if product names are blank

MKacprzakSolvd committed:
added support for other IDE to gitignore (from carina-demo)

MKacprzakSolvd committed:
replaced data provider with class that gives valid data

MKacprzakSolvd committed:
added test case desptiotions

MKacprzakSolvd committed:
added clearing shopping cart in verifyAddRemoveFromShoppingCartTest

MKacprzakSolvd committed:
Add missinng comment to assert, renamed method and added doc comment to it

MKacprzakSolvd committed:
Added missing message to assert

MKacprzakSolvd committed:
Improved way of specifying product category for ProductsPage and its loading strategy

MKacprzakSolvd committed:
Migrated from custom fixed carina-webdriver to official release

MKacprzakSolvd committed:
Created helepr function for measuring execution time

MKacprzakSolvd committed:
Added checking for selected size and color when testing filters

MKacprzakSolvd committed:
Improve checking if product matches one in the cart

Previously it just checked if product matches name of the item in the shopping cart, now it can also compare price.
I also created foundations for adding more checks in the future.
Also used new functionality in verifyAddRemoveFromShoppingCartTest.

MKacprzakSolvd committed:
Added missing access modifiers

MKacprzakSolvd committed:
Fixed calling Optional::get without checks

While isPresent was called before get, it was inside SoftAssert so it did nothing

MKacprzakSolvd committed:
Fixed typo

MKacprzakSolvd committed:
Fixed more typos

MKacprzakSolvd committed:
Changed timeMethod to accept Runnable instead of Supplier

whetman commented:
I think it would be good if you add loggers to your methods to improve readability and to be able to quicker resolve issues if any problem occurs

Trukhanovich commented on src/main/java/com/solvd/gui/pages/common/ProductsPageBase.java:
```diff
@@ -19,11 +19,17 @@
 
+    // this element is used as load marker, and > selector is used in order to
+    // distinguish between search and products page
+    @FindBy(css = ".main > .products.wrapper")
```
Could you please use more stable locator

MKacprzakSolvd replied:
It seems pretty stable to me, it's used as load marker, so it also needs to be uniqe. Thats why I'm using child selector (`>`) to distinguis it from SearchProductPage, which is very similar. Also, there doesn't seems to be any better way to locate this element, since there are no ids nearby. Also the classes don't seem likely to change their names.

<img width="650" alt="Screenshot 2024-06-28 at 13 23 04" src="https://github.com/MKacprzakSolvd/carina-automation-learning/assets/170347259/7e6230a5-f61f-4ada-9bcd-8657f0ad8dd1">

Trukhanovich commented on src/main/java/com/solvd/gui/pages/common/components/ProductCardBase.java:
```diff
@@ -53,24 +54,44 @@ public String getLink() {
+        return this.availableSizes.stream()
+                .filter(element -> element.getAttribute("aria-checked").equals("true"))
+                .map(element -> element.getAttribute("option-label"))
+                .findFirst();
```
why do you want to use findFirst() instead of findAny()

MKacprzakSolvd replied:
Sorry, my mistake. Resolved in df4c04ba568899d603b6cece0aea2d9f89e375ed

Trukhanovich commented on src/main/java/com/solvd/model/Product.java:
```diff
@@ -15,8 +16,55 @@ public class Product {
+        if (other == null) {
+            return true;
+        }
+        if (differsInValue(getName(), other.getName())) return true;
+        if (differsInValue(getPrice(), other.getPrice(), (a, b) -> a.compareTo(b) == 0)) return true;
+        if (differsInValue(getRating(), other.getRating())) return true;
+        if (differsInValue(getReviewNumber(), other.getReviewNumber())) return true;
+        if (differsInValue(getInStock(), other.getInStock())) return true;
```
could you please avoid such big count of if()

MKacprzakSolvd replied:
Resolved in commit d7e8b1977c4216d47a46b008acd92320002294be

MKacprzakSolvd committed:
Replaces findFirst with findAny

MKacprzakSolvd committed:
Joined multiple ifs into one

MKacprzakSolvd commented:
> I think it would be good if you add loggers to your methods to improve readability and to be able to quicker resolve issues if any problem occurs

Thanks, that is a good idea, though I think that is beyond the scope of this bugfix. I created issue for this (#3) and will solve it later.

AnnaSolvd commented:
Looks good to me. I would only improve readability of declarations of web elements, for example: [here](https://github.com/MKacprzakSolvd/carina-automation-learning/blob/d7e8b1977c4216d47a46b008acd92320002294be/src/main/java/com/solvd/gui/pages/common/CheckoutPageStepOneBase.java#L16C23-L16C46)

MKacprzakSolvd commented:
> Looks good to me. I would only improve readability of declarations of web elements, for example: [here](https://github.com/MKacprzakSolvd/carina-automation-learning/blob/d7e8b1977c4216d47a46b008acd92320002294be/src/main/java/com/solvd/gui/pages/common/CheckoutPageStepOneBase.java#L16C23-L16C46)

MKacprzakSolvd committed:
Added more empty lines for readability

MKacprzakSolvd commented:
> Looks good to me. I would only improve readability of declarations of web elements, for example: [here](https://github.com/MKacprzakSolvd/carina-automation-learning/blob/d7e8b1977c4216d47a46b008acd92320002294be/src/main/java/com/solvd/gui/pages/common/CheckoutPageStepOneBase.java#L16C23-L16C46)

I added more empty lines. If everything is ok, please remember to click approve (in Files changed tab > review changes)

MariuszSolvd commented:
You need to review **all** locators in your project. I look at them and it would be too much to go for each of them but I will give you one example, if you will not understand, we can get call and then I will explain further. So looking at your ProductsPageBase(but it happens also in other classes, another example is CheckoutPageStepOneBase) you have this [line](https://github.com/MKacprzakSolvd/carina-automation-learning/blob/62750fd45a0eb58e5ce23699feab7b5e42206637/src/main/java/com/solvd/gui/pages/common/ProductsPageBase.java#L46) and you xpath looks like: /**/*[@id='layered-filter-block']//*[contains(@class,'filter-options-item')][.//*[contains(@class,'filter-options-title')][text()='Color']]**, 
I made same with just **//div[text()='Color']/parent::div**
I tested your code with my locator and it works with this as well(I run it 10 times and it failed just once so I would say your test is very stable, good job!)
So that's first thing, please try to make your selectors **Stable**, **Flexible** and **Supportive**
For an example CheckoutPageStepOneBase you use idexes in locators, Andrei explained us it's really bad approach(it's not Stable) 

But let's back to ProductsPageBase. You could do another approach to this, I notice you already have ProductFilterBase as AbstractUIObject. Why not use there a List<ExtendedWebElement> menuButtons with this locator /**/div[@id = 'narrow-by-list']/*** and now you have all buttons in one place, component, and you can iterate over it. Think about it :)

MariuszSolvd commented:
About ProductFilterBase, I think you can improve it by using SearchContext which Carina provides. It will help you to improve readability and encapsulate code. As I mentioned before you could first of all make in you ProductBasePage representation of menuButtons. You can use this simple xPath to do it: _/div[@id = 'narrow-by-list']/*_ and using SearchContext you can create an object that will represent these buttons let's call it MenuButton. So ProductPageBase would have list of List<MenuButton>, in menu button you could create another object for an example ButtonOption. You could use this xPath: _.//div[contains(@class, 'swatch-attribute-options')]/*_ to get all options. For an example when you pick Color you will have all colors as a ButtonOption. This way you would have access to all buttons in your menu not only Colors and Size as you currently have, and available options.

MariuszSolvd commented:
Another example of [xPath](https://github.com/MKacprzakSolvd/carina-automation-learning/blob/62750fd45a0eb58e5ce23699feab7b5e42206637/src/main/java/com/solvd/gui/pages/common/ProductsPageBase.java#L50)
This would be enough, even when you put something to cart.
_//div[@class = 'minicart-wrapper']_

MariuszSolvd commented:
[HERE](https://github.com/MKacprzakSolvd/carina-automation-learning/blob/62750fd45a0eb58e5ce23699feab7b5e42206637/src/main/java/com/solvd/gui/pages/common/ProductsPageBase.java#L83) I would say name of this method is miss leading. Calling it like getProductFilterBaseByProductFilter or something similar would be better. You call this method in few places and moreover after calling it you call another method on this object [example](https://github.com/MKacprzakSolvd/carina-automation-learning/blob/62750fd45a0eb58e5ce23699feab7b5e42206637/src/main/java/com/solvd/gui/pages/common/ProductsPageBase.java#L121)
Would be nicer to have information what object you get by it since you in same line call a method that exist in this method.

MariuszSolvd commented:
[Here](https://github.com/MKacprzakSolvd/carina-automation-learning/blob/62750fd45a0eb58e5ce23699feab7b5e42206637/src/main/java/com/solvd/gui/pages/common/components/ProductFilterBase.java#L28) use of key word this is ok, but [here](https://github.com/MKacprzakSolvd/carina-automation-learning/blob/62750fd45a0eb58e5ce23699feab7b5e42206637/src/main/java/com/solvd/gui/pages/common/components/ProductFilterBase.java#L57) seems redundant.

[Here](https://github.com/MKacprzakSolvd/carina-automation-learning/blob/62750fd45a0eb58e5ce23699feab7b5e42206637/src/main/java/com/solvd/gui/pages/common/ProductsPageBase.java#L110) her you can use method reference.

MariuszSolvd commented:
Again, I am pretty sure these two lines has redundant this key word. [Here](https://github.com/MKacprzakSolvd/carina-automation-learning/blob/62750fd45a0eb58e5ce23699feab7b5e42206637/src/main/java/com/solvd/gui/pages/common/ProductsPageBase.java#L85) [Here](https://github.com/MKacprzakSolvd/carina-automation-learning/blob/62750fd45a0eb58e5ce23699feab7b5e42206637/src/main/java/com/solvd/gui/pages/common/ProductsPageBase.java#L86)
Pay attention to it, please.

MariuszSolvd commented:
I think [here](https://github.com/MKacprzakSolvd/carina-automation-learning/blob/62750fd45a0eb58e5ce23699feab7b5e42206637/src/main/java/com/solvd/gui/pages/common/ProductsPageBase.java#L25) another unstable selector. I made you more stable one _//ol[contains(@class, 'products')]/*_ in my opinion. It use just contains and is anchored on parent div. So even if dev team will decide to change something in class it's bigger probability mine will still works since it takes just contains for product not all class as your. Moreover when you will try to simplify yours to .product-item which would be nice approach it will get 13 results instead of correct 12, so I guess can't be use, in this case.

MariuszSolvd commented:
[Here](https://github.com/MKacprzakSolvd/carina-automation-learning/blob/62750fd45a0eb58e5ce23699feab7b5e42206637/src/main/java/com/solvd/gui/pages/common/components/ProductCardBase.java#L25) why put [@option-label] when in DOM you can see it has only one direct child you can easy replace it with /*

Moreover again in this class(ProductCardBase) I would use Nested UiObject way since you have n of Sizes why not make Object Size which will be AbstractUIObject and then there you can extract all needed data easier. And in your main class instead of using List<ExtendedWebElement> use List<Sizes>. It helps with readability, and maintenance of your code.

MKacprzakSolvd committed:
Improved locators

- Simplified them for CheckoutPageStepOneBase
- Changed css to className for ShoppingCartBase
- Improved stablitty for ProductsPageBase
- Improved stability for ProductDetailsPageBase

MKacprzakSolvd committed:
Renamed getFilter method to make code more readable

MKacprzakSolvd committed:
Further renamed getFilter method to make code more readable

MKacprzakSolvd committed:
Renamed variable to have different name from field

MKacprzakSolvd committed:
Replaced lambdas with method references where possible

MKacprzakSolvd commented:
Thank you very much for the review, here is my response to your comments:
1. [MKacprzakSolvd/carina-automation-learning#1 Bugfix/from meeting 2024 06 21 by...](https://github.com/MKacprzakSolvd/carina-automation-learning/pull/1#issuecomment-2198049052)
    > You need to review all locators in your project. I look at them and it would be too much to go for each of them but I will give you one example, if you will not understand, we can get call and then I will explain further. So looking at your ProductsPageBase(but it happens also in other classes, another example is CheckoutPageStepOneBase) you have this [line](https://github.com/MKacprzakSolvd/carina-automation-learning/blob/62750fd45a0eb58e5ce23699feab7b5e42206637/src/main/java/com/solvd/gui/pages/common/ProductsPageBase.java#L46) and you xpath looks like: //[@id='layered-filter-block']//[contains(@Class,'filter-options-item')][.//*[contains(@Class,'filter-options-title')][text()='Color']],
I made same with just //div[text()='Color']/parent::div
I tested your code with my locator and it works with this as well(I run it 10 times and it failed just once so I would say your test is very stable, good job!)
So that's first thing, please try to make your selectors Stable, Flexible and Supportive
For an example CheckoutPageStepOneBase you use idexes in locators, Andrei explained us it's really bad approach(it's not Stable)

    *(1)* About locators: when doing them, I general try to firs find some general wrapper component, like header create locator for it and then inside it locator for specific element (unless there is really good locator for this element, like unique id), I think that doing it that way make my locators more Supportive, because even if it breaks, it's easier to fix it. I will admit that my approach could slightly decrease locator stability, but if the locator for the general wrapper component is choosen correctly, it shouldn't be too much of a decrease.
          Another thing that you mentioned often in your suggestions about locators is using direct descendant (`/`) of xpath. I really don't like it, I think that it decreases locator stability. It will break if any element is injected between two elements, so adding any wrapper or any small change in structure will break you test.
          In the case that you mentioned `"//*[@id='layered-filter-block']" + "//*[contains(@class,'filter-options-item')]" + "[.//*[contains(@class,'filter-options-title')][text()='Size']]"` is in my opinion much more supportive than `//div[text()='Color']/parent::div` and is worth the tradeoff in my opinion (id and used classes are unlikely to change and even if one of them changes it seems easy to fix, moreover it's much less likely to lead to a false positive). I went through all my locators and fixed ones that seems bad.

    > But let's back to ProductsPageBase. You could do another approach to this, I notice you already have ProductFilterBase as AbstractUIObject. Why not use there a List menuButtons with this locator //div[@id = 'narrow-by-list']/* and now you have all buttons in one place, component, and you can iterate over it. Think about it :)

    *(2)* About ProductFilterBase, using it as a list will require a lot of changes to that component, since different filters have different strucure inside and would require different locators. And since it currently works and these improvements aren't required by any tests, I think that fixing it is outside of the scope of this pull request. I created issue #5 for it, and will get back to it when all tests will be stable.
2. [MKacprzakSolvd/carina-automation-learning#1 Bugfix/from meeting 2024 06 21 by...](https://github.com/MKacprzakSolvd/carina-automation-learning/pull/1#issuecomment-2198081243)
    > About ProductFilterBase, I think you can improve it by using SearchContext which Carina provides. It will help you to improve readability and encapsulate code.

    *(1)* I don't know what you mean by "using SearchContext that Carina providese". Could you elaborate on it?
    > As I mentioned before you could first of all make in you ProductBasePage representation of menuButtons. You can use this simple xPath to do it: /div[@id = 'narrow-by-list']/* and using SearchContext you can create an object that will represent these buttons let's call it MenuButton. So ProductPageBase would have list of List, in menu button you could create another object for an example ButtonOption. You could use this xPath: .//div[contains(@Class, 'swatch-attribute-options')]/* to get all options. For an example when you pick Color you will have all colors as a ButtonOption. This way you would have access to all buttons in your menu not only Colors and Size as you currently have, and available options.

    *(2)* About locators and ProductFilterBase - look at 1.1 and 1.2
3. [MKacprzakSolvd/carina-automation-learning#1 Bugfix/from meeting 2024 06 21 by...](https://github.com/MKacprzakSolvd/carina-automation-learning/pull/1#issuecomment-2198088984)
    > Another example of [xPath](https://github.com/MKacprzakSolvd/carina-automation-learning/blob/62750fd45a0eb58e5ce23699feab7b5e42206637/src/main/java/com/solvd/gui/pages/common/ProductsPageBase.java#L50)
This would be enough, even when you put something to cart.
//div[@Class = 'minicart-wrapper']

    *(1)* Look at 1.1
5. [MKacprzakSolvd/carina-automation-learning#1 Bugfix/from meeting 2024 06 21 by...](https://github.com/MKacprzakSolvd/carina-automation-learning/pull/1#issuecomment-2198091379)
    > [HERE](https://github.com/MKacprzakSolvd/carina-automation-learning/blob/62750fd45a0eb58e5ce23699feab7b5e42206637/src/main/java/com/solvd/gui/pages/common/ProductsPageBase.java#L83) I would say name of this method is miss leading. Calling it like getProductFilterBaseByProductFilter or something similar would be better. You call this method in few places and moreover after calling it you call another method on this object [example](https://github.com/MKacprzakSolvd/carina-automation-learning/blob/62750fd45a0eb58e5ce23699feab7b5e42206637/src/main/java/com/solvd/gui/pages/common/ProductsPageBase.java#L121)
Would be nicer to have information what object you get by it since you in same line call a method that exist in this method.

    *(1)* Good idea, renamed this method, though I think that better signature would be: `getFilterComponent( productFilterType )` - it's shorter and in my opinion equally expressive. (commits [84a92cd](https://github.com/MKacprzakSolvd/carina-automation-learning/pull/1/commits/84a92cdf4c90446e5ad0e31c7ef2f749e49338b9) and [3489a69](https://github.com/MKacprzakSolvd/carina-automation-learning/pull/1/commits/3489a69deac847ff25b73da8b2c49d42669ff09a))
7. [MKacprzakSolvd/carina-automation-learning#1 Bugfix/from meeting 2024 06 21 by...](https://github.com/MKacprzakSolvd/carina-automation-learning/pull/1#issuecomment-2198093151)
    > [Here](https://github.com/MKacprzakSolvd/carina-automation-learning/blob/62750fd45a0eb58e5ce23699feab7b5e42206637/src/main/java/com/solvd/gui/pages/common/components/ProductFilterBase.java#L28) use of key word this is ok,

    *(1)* Actually this seems like errro on my part, fixed it. (commit [6f21185](https://github.com/MKacprzakSolvd/carina-automation-learning/pull/1/commits/6f21185d398f6ec840a7926210a4369630a12c50))


    >  but [here](https://github.com/MKacprzakSolvd/carina-automation-learning/blob/62750fd45a0eb58e5ce23699feab7b5e42206637/src/main/java/com/solvd/gui/pages/common/components/ProductFilterBase.java#L57) seems redundant. 

    *(2)* About this usage, for me using this for fields makes code more readable - it makes easier to distinguis between fields and local variables, but I read about it and majority of people seems to suggest avoiding it's usage. I won't fix it in this pull request, because it would create a lot of changes and would require resolving conflicts in another pull requests, but I created issue #6 to fix it later.
    >  [Here](https://github.com/MKacprzakSolvd/carina-automation-learning/blob/62750fd45a0eb58e5ce23699feab7b5e42206637/src/main/java/com/solvd/gui/pages/common/ProductsPageBase.java#L110) her you can use method reference.

    *(3)* Done (commit [0979c06](https://github.com/MKacprzakSolvd/carina-automation-learning/pull/1/commits/0979c0697fe6efb518941c61cb1c88394e02f139))
9. [MKacprzakSolvd/carina-automation-learning#1 Bugfix/from meeting 2024 06 21 by...](https://github.com/MKacprzakSolvd/carina-automation-learning/pull/1#issuecomment-2198098013)
    >  Again, I am pretty sure these two lines has redundant this key word. [Here](https://github.com/MKacprzakSolvd/carina-automation-learning/blob/62750fd45a0eb58e5ce23699feab7b5e42206637/src/main/java/com/solvd/gui/pages/common/ProductsPageBase.java#L85) [Here](https://github.com/MKacprzakSolvd/carina-automation-learning/blob/62750fd45a0eb58e5ce23699feab7b5e42206637/src/main/java/com/solvd/gui/pages/common/ProductsPageBase.java#L86)
Pay attention to it, please.

    *(1)* Same as 5.2
11. [MKacprzakSolvd/carina-automation-learning#1 Bugfix/from meeting 2024 06 21 by...](https://github.com/MKacprzakSolvd/carina-automation-learning/pull/1#issuecomment-2198106778)
    > I think [here](https://github.com/MKacprzakSolvd/carina-automation-learning/blob/62750fd45a0eb58e5ce23699feab7b5e42206637/src/main/java/com/solvd/gui/pages/common/ProductsPageBase.java#L25) another unstable selector. I made you more stable one //ol[contains(@Class, 'products')]/* in my opinion. It use just contains and is anchored on parent div. So even if dev team will decide to change something in class it's bigger probability mine will still works since it takes just contains for product not all class as your.

    *(1)* See 1.1 - I think it would be more stable to avoid ol, and using `/` also makes it less stable in my opinion
    > Moreover when you will try to simplify yours to .product-item which would be nice approach it will get 13 results instead of correct 12, so I guess can't be use, in this case.

    *(2)* About 13 results, could you post link to page on which you get 13 items? I checked few and always got 12.
13. [MKacprzakSolvd/carina-automation-learning#1 Bugfix/from meeting 2024 06 21 by...](https://github.com/MKacprzakSolvd/carina-automation-learning/pull/1#issuecomment-2198109940)
    > [Here](https://github.com/MKacprzakSolvd/carina-automation-learning/blob/62750fd45a0eb58e5ce23699feab7b5e42206637/src/main/java/com/solvd/gui/pages/common/components/ProductCardBase.java#L25) why put [@option-label] when in DOM you can see it has only one direct child you can easy replace it with /*

    *(1)* See 1.1 about `@option-label` vs `/`
    > Moreover again in this class(ProductCardBase) I would use Nested UiObject way since you have n of Sizes why not make Object Size which will be AbstractUIObject and then there you can extract all needed data easier. And in your main class instead of using List use List. It helps with readability, and maintenance of your code.

    *(2)* Good idea, but outside of the scope of this pull request, created issue #7, will do it later, thanks

AnnaSolvd commented:
LGTM , but there's no need to comment general things. For example, method name is getRandomElement so you know what this line do [here](https://github.com/MKacprzakSolvd/carina-automation-learning/blob/0979c0697fe6efb518941c61cb1c88394e02f139/src/test/java/com/solvd/ContextSwitchingTest.java#L54)

whetman reviewed (approved):
I don't see any issues, but I agree with Ania about the comments

MKacprzakSolvd committed:
improved comments

made them more consistent, deleted unneeded

MKacprzakSolvd committed:
Typos

MKacprzakSolvd committed:
Fixed running tests in parallel for different platforms

Now it loads properties just for current test, instead of loading them globally
